server/util/proto/proto.go

```diff
@@ -41,3 +41,11 @@ func Unmarshal(b []byte, v Message) error {
 
 	return UnmarshalOld(b, v)
 }
+
+func Clone(v Message) Message {
+	vt, ok := v.(vtprotoMessage)
+	if ok {
+		return vt.CloneMessageVT()
+	}
+	return gproto.Clone(v)
```

Comment:
should we stick with the `<Method>` + `<Method>Old` convention and have separate `Clone` for VTProto, `CloneOld` for old-style protos?

Reply:
To clarify, `Clone()` (for VT-only) could look like this:

```
type typedVtprotoMessageVT[T any] interface {
  CloneVT() T
}

func Clone[T any](msg typedVTProtoMessage[T]) T {
  return msg.CloneVT()
}
```

Reply:
I want to keep the method signature of `proto.Clone` the same as `gproto.Clone`. This can be used in the case when we clone a generic `proto.Message` and still want to utilized the speedup from vtproto-generated code.  One example of this is in [auditlog.go](https://github.com/buildbuddy-io/buildbuddy/blob/bd8946dac8f6b3cd979daa28151eee66e4c7a1e4/enterprise/server/auditlog/auditlog.go#L4). I don't think we want to use proto.CloneOld in this case, because CloneOld (i.e. the reflection-based Clone method is slower). 

We also have cases where we are cloning a message without vtproto generated code because they are defined outside buildbuddy repository. e.g. in [prom.go]((https://github.com/buildbuddy-io/buildbuddy/blob/bd8946dac8f6b3cd979daa28151eee66e4c7a1e4/enterprise/server/backends/prom/prom.go#L327)), we clone `dto.MetricFamily`. In this case, the proto.Clone method defined here also works; and it will use gproto.Clone. (The caller doesn't need to explicitly call proto.CloneOld) 

In the case where we know the type of proto messages, it's *not* necessary to use `CloneVT()` at all. I updated the callsites in https://github.com/buildbuddy-io/buildbuddy/pull/5651 because I felt it's cleaner to use CloneVT() and can also avoid unnecessary type casting.
There are three options in this case:
1. keep the callsites as it is:
```
newMessage := proto.Clone(oldMessage).(*protoType)
```
2. Use CloneVT()
```
newMessage := oldMessage.CloneVT()
```
3. Implement a new method CloneTyped() using generic interface as you suggested:
```
newMessage := proto.CloneTyped[*proto.Type](oldMessage)
```

Personally, I like option 2 better (though there is VT suffix) for aesthetic reasons, but I can also do option 3 if people feel strongly about it.

Reply:
In option 3, the `[*proto.Type]` part can be omitted because the type can be inferred, so it would look like `newMessage := proto.CloneTyped(oldMessage)`. IMO it's slightly clearer than `CloneVT()` but I don't feel super strongly about it.

Reply:
Oh..I assumed I couldn't do that b/c my editor reported an error when I don't have the type specified.
```
type *raft.FileMetadata of md does not match "github.com/buildbuddy-io/buildbuddy/server/util/proto".typedVTProtoMessage[T] (cannot infer T) (type *raft.FileMetadata of md does not match)
```
But now you mentioned it, I ran the test with that code and it seemed fine.